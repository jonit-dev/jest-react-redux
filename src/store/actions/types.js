export const ADD_GIFT = "ADD_GIFT";
export const DELETE_GIFT = "DELETE_GIFT";
export const FETCH_POSTS = "FETCH_POSTS";
