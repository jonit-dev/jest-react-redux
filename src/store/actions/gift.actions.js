import { ADD_GIFT, DELETE_GIFT } from "./types";

export const addGift = gift => {
  return { type: ADD_GIFT, payload: gift };
};

export const deleteGift = id => {
  return { type: DELETE_GIFT, payload: id };
};
