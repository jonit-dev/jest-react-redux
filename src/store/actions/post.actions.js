import axios from "axios";
import { FETCH_POSTS } from "./types";
export const getPosts = () => async dispatch => {
  const response = await axios.get(
    "https://jsonplaceholder.typicode.com/posts"
  );

  dispatch({ type: FETCH_POSTS, payload: response.data });
};
