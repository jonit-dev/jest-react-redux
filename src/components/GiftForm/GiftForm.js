import React, { Component } from "react";
import ListItem from "../List/ListItem";
import { connect } from "react-redux";
import { addGift, deleteGift } from "../../store/actions/gift.actions";
import { getPosts } from "../../store/actions/post.actions";

export class GiftForm extends Component {
  state = {
    gift: {
      description: "",
      person: ""
    }
  };

  componentDidMount() {
    this.props.getPosts().then(posts => {
      console.log(this.props.posts);
    });
  }

  onClearGiftInput() {
    this.setState({
      gift: {
        description: "",
        person: ""
      }
    });
  }

  onRenderGiftList() {
    return this.props.gifts.map(gift => (
      <ListItem
        {...gift}
        key={gift.id}
        onDeleteItem={id => this.props.deleteGift(id)}
      />
    ));
  }

  onAddNewGift() {
    //check if gift to be added is not null

    if (!this.state.gift.person || !this.state.gift.description) {
      alert("You cannot add an empty gift");
      return false;
    }

    //add new gift to the redux store
    this.props.addGift({
      id: this.props.gifts.length + 1,
      ...this.state.gift
    });

    // clear local state

    this.onClearGiftInput();
  }

  render() {
    return (
      <>
        <form>
          <div className="form-group">
            <label htmlFor="exampleInputEmail1">Person</label>
            <input
              type="text"
              className="form-control input-person"
              placeholder="Enter a new gift"
              value={this.state.gift.person}
              onChange={e =>
                this.setState({
                  gift: { ...this.state.gift, person: e.target.value }
                })
              }
            />
          </div>
          <div className="form-group">
            <label htmlFor="exampleInputEmail1">Gift</label>
            <input
              type="text"
              className="form-control input-gift"
              placeholder="Enter a new gift"
              value={this.state.gift.description}
              onChange={e =>
                this.setState({
                  gift: {
                    ...this.state.gift,
                    description: e.target.value
                  }
                })
              }
            />
          </div>
          <button
            type="button"
            className="btn btn-primary btn-lg btn-block btn-add"
            onClick={() => this.onAddNewGift()}
          >
            Add a new gift
          </button>
        </form>

        <hr />

        <ul className="list-group gift-list">{this.onRenderGiftList()}</ul>
      </>
    );
  }
}
const mapStateToProps = state => {
  return { gifts: state.giftsReducer.gifts, posts: state.postsReducer.posts };
};

export default connect(
  mapStateToProps,
  {
    addGift,
    deleteGift,
    getPosts
  }
)(GiftForm);
