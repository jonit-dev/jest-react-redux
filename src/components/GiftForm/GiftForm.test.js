import React from "react";
import { shallow } from "enzyme";
import { GiftForm } from "./GiftForm";
import { ADD_GIFT, DELETE_GIFT, FETCH_POSTS } from "../../store/actions/types";
import { addGift, deleteGift } from "../../store/actions/gift.actions";
import giftReducer from "../../store/reducers/gift.reducer";

import { getPosts } from "../../store/actions/post.actions";
import mockStore from "../../store/reduxMockStore";
import fetchMock from "fetch-mock";

//setup the gift shallow
const mockRemove = jest.fn();
const id = 1;
const props = {
  gifts: [{ id: 1, description: "hello world", person: "myself" }],
  gift: {
    description: "",
    person: ""
  },
  removeGift: mockRemove,
  getPosts: getPosts(),
  posts: []
};

const giftForm = shallow(<GiftForm {...props} />);

/*#############################################################|
|  >>> DEFAULT REACT COMPONENT TESTING
*##############################################################*/

describe("Make sure component is bootstrapping correctly", () => {
  // Initialize mockstore with empty state

  it("renders the component correctly", () => {
    expect(giftForm).toMatchSnapshot();
  });

  it("initializes state correctly", () => {
    expect(giftForm.state().gift).toEqual({
      description: "",
      person: ""
    });
  });
});

/*#############################################################|
 |  >>> REDUX
 *##############################################################*/

const gift = {
  id: 2,
  description: "second gift",
  person: "chuck norris"
};

// Reducers ========================================

describe("Reducers: giftReducer", () => {
  it("should return default state", () => {
    const INITIAL_STATE = giftReducer(undefined, {});
    expect(INITIAL_STATE).toEqual({
      gifts: [{ id: 1, description: "hello world", person: "myself" }]
    });
  });

  it("should add a new gift if addGift action is triggered", () => {
    const updatedState = giftReducer(undefined, {
      type: ADD_GIFT,
      payload: gift
    });

    expect(updatedState).toEqual({
      gifts: [
        { id: 1, description: "hello world", person: "myself" },
        { id: 2, description: "second gift", person: "chuck norris" }
      ]
    });
  });

  it("should remove gift when deleteGift is triggered", () => {
    const id = 1;

    const updatedState = giftReducer(undefined, {
      type: DELETE_GIFT,
      payload: id
    });

    expect(updatedState).toEqual({
      gifts: []
    });
  });
});

// Actions (sync) ========================================

describe("Actions: Synchronous gift actions working properly", () => {
  it("should create an action to add a gift", () => {
    const expectedAction = { type: ADD_GIFT, payload: gift };

    expect(addGift(gift)).toEqual(expectedAction);
  });

  it("should create an action to remove a gift", () => {
    const expectedAction = { type: DELETE_GIFT, payload: 1 };

    expect(deleteGift(1)).toEqual(expectedAction);
  });
});

// Actions (async) ========================================

// describe("Actions: Asynchronous gift actions should work properly", () => {
//   afterEach(() => {
//     fetchMock.restore();
//   });

//   it("calls getPosts and update store successfully", () => {
//     const posts = [
//       {
//         userId: 1,
//         id: 1,
//         title:
//           "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
//         body:
//           "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
//       },
//       {
//         userId: 1,
//         id: 2,
//         title: "qui est esse",
//         body:
//           "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
//       }
//     ];

//     fetchMock.getOnce("https://jsonplaceholder.typicode.com/posts", {
//       body: {
//         data: posts
//       },
//       headers: { "content-type": "application/json" }
//     });

//     const expectedActions = [{ type: FETCH_POSTS, payload: posts }];

//     const store = mockStore({ posts: [] });

//     return store.dispatch(getPosts()).then(() => {
//       console.log(store.getActions());
//       const actions = store.getActions();

//       expect(actions).toBe(expectedActions);
//     });
//   });
// });
