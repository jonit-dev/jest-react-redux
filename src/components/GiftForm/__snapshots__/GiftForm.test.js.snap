// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make sure component is bootstrapping correctly renders the component correctly 1`] = `ShallowWrapper {}`;
